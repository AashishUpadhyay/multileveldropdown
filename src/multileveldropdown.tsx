import React, { useState } from "react";
import { withStyles } from "@material-ui/core/styles";
import {
  List,
  ListItem,
  ListItemText,
  Menu,
  MenuItem
} from "@material-ui/core";

const styles = () => ({
  boldListItem: {
    fontWeight: "bold" as "bold"
  },
  nestedListItem: {
    paddingLeft: "2rem"
  }
});

export interface IMultilevelDropdownProps {
  label: string;
  options: IMultilevelDropdownOption[];
  selectedId: number;
  classes?: any;
  setOption: (option: IMultilevelDropdownOption) => void;
}

export interface IMultilevelDropdownOption {
  Id: number;
  Name: string;
  Type: string;
  Children: IMultilevelDropdownOption[];
}

export const findSelectedOption = (
  options: IMultilevelDropdownOption[],
  id: number
) => {
  const selected = options.filter((option: IMultilevelDropdownOption) =>
    option.Children.some((child: IMultilevelDropdownOption) => child.Id === id)
  );
  if (selected[0].Children.length === 1) {
    return selected[0].Name;
  }
  return selected.map((option: any) =>
    option.Children.find((child: IMultilevelDropdownOption) => child.Id === id)
  )[0].Name;
};

export const composeOptionMenu = (
  options: IMultilevelDropdownOption[],
  id: number,
  boldListItem: any,
  nestedListItem: any,
  handleMenuClick: (option: IMultilevelDropdownOption) => void
) => {
  return options.map((option: IMultilevelDropdownOption) => {
    const isHeaderOption = option.Id <= 0;
    const isLeveled = option.Children.length > 1;
    const menuItems = [];

    menuItems.push(
      <MenuItem
        key={isHeaderOption ? option.Children[0].Id : option.Id}
        className={isHeaderOption || isLeveled ? boldListItem : undefined}
        selected={
          !isHeaderOption && !isLeveled
            ? option.Children[0].Id === id
            : undefined
        }
        onClick={
          !isHeaderOption && !isLeveled
            ? () => handleMenuClick(option.Children[0])
            : undefined
        }
      >
        {option.Name}
      </MenuItem>
    );
    if (isLeveled) {
      option.Children.forEach((child: IMultilevelDropdownOption) =>
        menuItems.push(
          <MenuItem
            key={child.Id}
            selected={child.Id === id}
            className={nestedListItem}
            onClick={(e: React.MouseEvent) => handleMenuClick(child)}
          >
            {child.Name}
          </MenuItem>
        )
      );
    }
    return menuItems;
  });
};

const BaseMultilevelDropdown: React.FunctionComponent<
  IMultilevelDropdownProps
> = ({ classes: { boldListItem, nestedListItem }, ...props }) => {

  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const {
    label,
    options,
    selectedId,
    setOption
  } = props;

  const handleMenuClick = (option: IMultilevelDropdownOption) => {
    setOption(option);
    setAnchorEl(null);
  }

  const handleClickListItem = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget);
  }

  const handleClose = () => {
    setAnchorEl(null);
  }

  return (
    <div>
      <List component="nav" aria-label={label}>
        <ListItem
          button
          aria-haspopup="true"
          aria-controls="lock-menu"
          aria-label={label}
          onClick={e => handleClickListItem(e)}
        >
          <ListItemText
            primary={label}
            secondary={findSelectedOption(options, selectedId)}
          />
        </ListItem>
      </List>
      <Menu
        id="lock-menu"
        anchorEl={anchorEl}
        keepMounted
        open={Boolean(anchorEl)}
        onClose={() => handleClose()}
      >
        {composeOptionMenu(
          options,
          selectedId,
          boldListItem,
          nestedListItem,
          handleMenuClick
        )}
      </Menu>
    </div>
  );
};

const MultilevelDropdown = withStyles(styles)(BaseMultilevelDropdown);

export { MultilevelDropdown };
