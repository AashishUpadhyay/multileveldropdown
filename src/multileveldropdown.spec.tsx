import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import * as React from 'react';
import { findSelectedOption, composeOptionMenu, MultilevelDropdown, IMultilevelDropdownProps, IMultilevelDropdownOption } from './multileveldropdown';

describe('multileveldropdown tests', () => {
  let options: IMultilevelDropdownOption[];
  let optionsWithHeader: IMultilevelDropdownOption[];

  beforeEach(() => {
    configure({ adapter: new Adapter() });

    options = [
      {
        Id: 17,
        Name: 'Vulnerability Library',
        Type: 'Application',
        Children: [
          {
            Id: 126,
            Name: 'Vulnerability Library',
            Type: 'Application',
            Children: []
          }
        ]
      },
      {
        Id: 65,
        Name: 'Policies',
        Type: 'Application',
        Children: [
          {
            Id: 3,
            Name: 'Policy',
            Type: 'Application',
            Children: []
          },
          {
            Id: 4,
            Name: 'Area',
            Type: 'Application',
            Children: []
          },
          {
            Id: 5,
            Name: 'Section',
            Type: 'Application',
            Children: []
          }
        ]
      },
      {
        Id: 66,
        Name: 'Control Standards',
        Type: 'Application',
        Children: [
          {
            Id: 41,
            Name: 'Control Standards',
            Type: 'Application',
            Children: []
          }
        ]
      },
      {
        Id: 537,
        Name: 'ARCHER-68342',
        Type: 'Application',
        Children: [
          {
            Id: 335,
            Name: 'ARCHER-68342',
            Type: 'Application',
            Children: []
          }
        ]
      },
      {
        Id: 539,
        Name: 'Containers',
        Type: 'Application',
        Children: [
          {
            Id: 337,
            Name: 'Containers',
            Type: 'Application',
            Children: []
          }
        ]
      }
    ];

    optionsWithHeader = [
      {
        Id: 17,
        Name: 'Vulnerability Library',
        Type: 'Application',
        Children: [
          {
            Id: 126,
            Name: 'Vulnerability Library',
            Type: 'Application',
            Children: []
          }
        ]
      },
      {
        Id: 65,
        Name: 'Policies',
        Type: 'Application',
        Children: [
          {
            Id: 3,
            Name: 'Policy',
            Type: 'Application',
            Children: []
          },
          {
            Id: 4,
            Name: 'Area',
            Type: 'Application',
            Children: []
          },
          {
            Id: 5,
            Name: 'Section',
            Type: 'Application',
            Children: []
          }
        ]
      },
      {
        Id: 66,
        Name: 'Control Standards',
        Type: 'Application',
        Children: [
          {
            Id: 41,
            Name: 'Control Standards',
            Type: 'Application',
            Children: []
          }
        ]
      },
      {
        Id: 537,
        Name: 'ARCHER-68342',
        Type: 'Application',
        Children: [
          {
            Id: 335,
            Name: 'ARCHER-68342',
            Type: 'Application',
            Children: []
          }
        ]
      },
      {
        Id: 539,
        Name: 'Containers',
        Type: 'Application',
        Children: [
          {
            Id: 337,
            Name: 'Containers',
            Type: 'Application',
            Children: []
          }
        ]
      },
      { Type: '0', Name: '42', Id: 0, Children: [{ Name: '42', Id: 0, Type: '', Children: [] }] },
      { Type: 'a', Name: '42', Id: -1, Children: [{ Name: '42', Id: -1, Type: '', Children: [] }] },
      { Type: 'o', Name: '42', Id: -2, Children: [{ Name: '42', Id: -2, Type: '', Children: [] }] }
    ];
  });

  it('should match snapshot', () => {
    const props: IMultilevelDropdownProps = {
      label: 'Target',
      options,
      selectedId: 126,
      setOption: (option: IMultilevelDropdownOption) => { }
    };
    const componentmount = mount(<MultilevelDropdown {...props} />);
    expect(componentmount).toMatchSnapshot();
  });

  it('should findSelectedOption for multilevel options', () => {
    const expectedName = 'Section';
    const returnedValue = findSelectedOption(options, 5);
    expect(returnedValue).toStrictEqual(expectedName);
  });

  it('should findSelectedOption for single level options', () => {
    const expectedName = 'Vulnerability Library';
    const returnedValue = findSelectedOption(options, 126);
    expect(returnedValue).toStrictEqual(expectedName);
  });

  it('should findSelectedOption for multilevel options when header options are present', () => {
    const expectedName = 'Section';
    const returnedValue = findSelectedOption(optionsWithHeader, 5);
    expect(returnedValue).toStrictEqual(expectedName);
  });

  it('should findSelectedOption for single level options when header options are present', () => {
    const expectedName = 'Vulnerability Library';
    const returnedValue = findSelectedOption(optionsWithHeader, 126);
    expect(returnedValue).toStrictEqual(expectedName);
  });

  it('should composeOptionMenu for single and multilevel options', () => {
    const expected = [17, 65, 3, 4, 5, 66, 537, 539];
    const boldListItem = {
      fontWeight: 'bold' as 'bold'
    };
    const nestedListItem = {
      paddingLeft: '2rem'
    };
    const setSelectedTarget = jest.fn();
    const returnedValue = composeOptionMenu(options, 126, boldListItem, nestedListItem, setSelectedTarget);
    const returnedValueFlattened = returnedValue.reduce((accumulator, value) => accumulator.concat(value), [] as JSX.Element[]);
    const returnedKeys = returnedValueFlattened.map(menuItem => {
      return menuItem.key;
    });

    expect(expected.join()).toStrictEqual(returnedKeys.join());
  });

  it('should composeOptionMenu for single and multilevel options when header options are present', () => {
    const expected = [17, 65, 3, 4, 5, 66, 537, 539, 0, -1, -2];
    const boldListItem = {
      fontWeight: 'bold' as 'bold'
    };
    const nestedListItem = {
      paddingLeft: '2rem'
    };
    const setSelectedTarget = jest.fn();
    const returnedValue = composeOptionMenu(optionsWithHeader, 126, boldListItem, nestedListItem, setSelectedTarget);
    const returnedValueFlattened = returnedValue.reduce((accumulator, value) => accumulator.concat(value), [] as JSX.Element[]);
    const returnedKeys = returnedValueFlattened.map(menuItem => {
      return menuItem.key;
    });

    expect(expected.join()).toStrictEqual(returnedKeys.join());
  });
});
