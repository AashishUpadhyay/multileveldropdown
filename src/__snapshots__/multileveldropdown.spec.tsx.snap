// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multileveldropdown tests should match snapshot 1`] = `ReactWrapper {}`;
